// @nx/react/src/migrations/update-14-0-0/add-default-development-configurations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.update = void 0;
const devkit_1 = require("@nx/devkit");
async function update(tree) {
    const projects = (0, devkit_1.getProjects)(tree);
    projects.forEach((config, name) => {
        let shouldUpdate = false;
        if (config.targets?.build?.executor === '@nrwl/web:webpack') {
            shouldUpdate = true;
            config.targets.build.defaultConfiguration ??= 'production';
            config.targets.build.configurations ??= {};
            config.targets.build.configurations.development ??= {
                extractLicenses: false,
                optimization: false,
                sourceMap: true,
                vendorChunk: true,
            };
        }
        if (config.targets?.serve?.executor === '@nrwl/web:dev-server') {
            shouldUpdate = true;
            config.targets.serve.defaultConfiguration ??= 'development';
            config.targets.serve.configurations ??= {};
            config.targets.serve.configurations.development ??= {
                buildTarget: `${name}:build:development`,
            };
        }
        if (shouldUpdate)
            (0, devkit_1.updateProjectConfiguration)(tree, name, config);
    });
    await (0, devkit_1.formatFiles)(tree);
}
exports.update = update;
exports.default = update;
