// @nx/react/src/migrations/update-14-0-0/update-react-dom-render-for-v18.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.migrateReactDomRender = exports.update = void 0;
const devkit_1 = require("@nx/devkit");
const js_1 = require("@nx/js");
const ts = require("typescript");
async function update(tree) {
    const projects = (0, devkit_1.getProjects)(tree);
    projects.forEach((config, name) => {
        const isReactProject = config.targets?.build?.executor === '@nrwl/web:webpack' &&
            /main\.(t|j)sx?$/.test(config.targets.build.options.main);
        if (isReactProject) {
            const sourcePath = config.targets.build.options.main;
            const sourceCode = tree.read(sourcePath).toString();
            const source = ts.createSourceFile(sourcePath, sourceCode, ts.ScriptTarget.Latest, true);
            const result = (0, devkit_1.applyChangesToString)(sourceCode, migrateReactDomRender(sourcePath, source));
            tree.write(sourcePath, result);
        }
    });
    await (0, devkit_1.formatFiles)(tree);
}
exports.update = update;
function migrateReactDomRender(sourcePath, source) {
    const allImports = (0, js_1.findNodes)(source, ts.SyntaxKind.ImportDeclaration);
    const reactDomImport = allImports.find((x) => x.moduleSpecifier.getText() === "'react-dom'");
    const changes = [];
    if (reactDomImport) {
        changes.push({
            type: devkit_1.ChangeType.Insert,
            index: reactDomImport.moduleSpecifier.end - 1,
            text: '/client',
        });
    }
    const calls = (0, js_1.findNodes)(source, ts.SyntaxKind.CallExpression);
    const renderCall = calls.find((x) => {
        if (x.expression.kind !== ts.SyntaxKind.PropertyAccessExpression)
            return false;
        const expr = x.expression;
        return (expr.expression.getText() === 'ReactDOM' &&
            expr.name.getText() === 'render');
    });
    if (renderCall) {
        const [element, querySelector] = renderCall.arguments;
        changes.push({
            type: devkit_1.ChangeType.Delete,
            start: renderCall.getStart(),
            length: renderCall.end,
        }, {
            type: devkit_1.ChangeType.Insert,
            index: renderCall.getStart(),
            text: (0, devkit_1.stripIndents) `
          const root = ReactDOM.createRoot(
            ${querySelector.getText()}${sourcePath.endsWith('.tsx') ? ' as HTMLElement' : ''}
          );
          root.render(
            ${element.getText()}
          );
        `,
        });
    }
    return changes;
}
exports.migrateReactDomRender = migrateReactDomRender;
exports.default = update;
